xtra-libp2p/src/endpoint.rs

```diff
@@ -526,26 +529,53 @@ impl Endpoint {
             "Type-system enforces that we only try to negotiate one protocol"
         );
 
-        let (protocol, stream) = self.open_substream(peer, protocols.clone()).await?;
+        let (control, _) = self.controls.get(&peer).ok_or(Error::NoConnection(peer))?;
 
-        debug_assert!(
-            protocol == protocols[0],
-            "If negotiation is successful, must have selected the only protocol we sent."
-        );
+        let fut = {
+            let connection_timeout = self.connection_timeout;
+            let control = control.clone();
+            async move {
+                let (protocol, stream) =
+                    Self::open_substream(control, peer, protocols.clone(), connection_timeout)
+                        .await?;
 
-        Ok(stream)
+                debug_assert!(
+                    protocol == protocols[0],
+                    "If negotiation is successful, must have selected the only protocol we sent."
+                );
+
+                Ok(stream)
+            }
+        };
+
+        Ok(Box::pin(fut))
     }
 
+    #[must_use]
     async fn handle(
         &mut self,
         msg: OpenSubstream<Multiple>,
-    ) -> Result<(&'static str, Substream), Error> {
+    ) -> Result<
+        Pin<Box<dyn futures::Future<Output = Result<(&'static str, Substream), Error>> + Send>>,
+        Error,
+    > {
         let peer = msg.peer;
         let protocols = msg.protocols;
 
-        let (protocol, stream) = self.open_substream(peer, protocols).await?;
+        let (control, _) = self.controls.get(&peer).ok_or(Error::NoConnection(peer))?;
+
+        let fut = {
+            let connection_timeout = self.connection_timeout;
+            let control = control.clone();
+            async move {
+                let (protocol, stream) =
+                    Self::open_substream(control, peer, protocols, connection_timeout).await?;
+
+                Ok((protocol, stream))
+            }
+        };
 
-        Ok((protocol, stream))
+        Ok(Box::pin(fut))
```

Comment:
what's the upside of returning a `Box::pin(fut)` instead of oneshot receiver here?

Reply:
I think it's slightly more direct. Returning a oneshot would look like

```rust
let (tx, rx) = oneshot();
tokio::spawn(fut.map(move |r| tx.send(r)));
Ok(rx)
```

This way has one less task spawned

Reply:
Not sure. Just what first came to mind when we identified the problem.

It seems more direct to just return a future as the result of the handler, rather than having to spawn a task, pass in a sender and return a receiver. But I'm not sure what's better in terms of performance.

Reply:
I would only expect microseconds difference here, and since this is a network operation it probably doesn't matter (?) as that will usually take milliseconds and more

Reply:
it would be good to see a performance comparison, might run them both with the load test when they're here to see. 

this is likely going to be our main bottleneck as we scale (it is already :P) , so picking up faster is likely to make a difference. However, I have no idea which one will be faster without looking at some results.

Reply:
The only performance gain I could see is that if you spawn the task, you start it before the receiving task is woken, which could save a few microseconds, and then you get the finishing the future -> receive anyway to pay that spawn back. So, I don't think it should be faster, but a benchmark could be nice. I'm not sure it'd be measurable in release, though.

Reply:
fair enough! so I guess `oneshot` would only make sense if the future would not be `Send`?

Reply:
also: how does the memory footprint change? is pinning only passing already allocated memory on the heap, or are we doing new allocations?

Reply:
> fair enough! so I guess oneshot would only make sense if the future would not be Send?

I think it would need to be Send to spawn anyway

> also: how does the memory footprint change? is pinning only passing already allocated memory on the heap, or are we doing new allocations?

`Box::pin` does allocate new memory, yes, but I think that tokio might do this under the hood with `spawn` anyway? I'm not sure - maybe it's more optimised...